### PR TITLE
recommend homebrew upgrade when installed with homebrew

### PR DESCRIPTION
**Pull request by akupila**
- try to determine if wercker was installed with homebrew
- if yes and a new version is available recommend user to use `brew upgrade wercker-cli` instead of the built in upgrade mechanism

### CODE REVIEW

**Comment by faiq** [comment]
I think the idea is good here, but ultimately i feel 👎  about this pull request and heres why: 

1. Theres an uncomfortable amount of `exec` usage. I know that we have to use it sometimes, but we should limit executing commands on a another person's machine. 
2. A lot of string parsing. Homebrew commands and command outputs are subject to change. Relying on this is not a good idea. 

Overall, I don't feel confident that this code is very robust and it solves a little problem. Most people know how they installed wercker, so I'm assuming they can figure out that they can update it as well.

**Comment by hatchan** [comment]
I don't see anything wrong with using exec, that is the only way to interact with the API of brew. 

I do agree that we shouldn't use string parsing. There is the JSON output which I think should stay stable (since it includes a version in the flag). So I think we should invoke the command, and parse the output as JSON. https://github.com/Homebrew/brew/blob/master/share/doc/homebrew/Querying-Brew.md

File: util/homebrew.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+//
+// Heuristic (bail if any step fails):
+//   1. see if homebrew is installed
+//   2. get installed packages from homebrew, look for `wercker-cli`
+//   3. find symlink installed by homebrew, follow it to binary
+//   4. get path of currently running executable
+//   5. compare path with homebrew path
+func InstalledWithHomebrew() (bool, error) {
+	// Check if homebrew is installed
+	cmd := exec.Command("which", "brew")
```
Comment (hatchan) [inline util/homebrew.go]: Don't think we need to do this. We can just try invoking brew, and if it fails just return `bool=false`

**Comment by hatchan** [comment]
We can consider placing all code related to the brew JSON parsing in a opensource repo (ie. the structs)

File: cmd/main.go [hunk 2]
```diff
@@ -770,6 +771,19 @@ func cmdVersion(options *core.VersionOptions) error {
 		if updater.UpdateAvailable() {
 			logger.Infoln("A new version is available:",
 				updater.ServerVersion.FullVersion())
+
+			// try to determine if binary was installed with homebrew
+			// ignore potential error; an error would mean we can't be sure
+			// so optimize for happy-path
+			homebrew, _ := util.InstalledWithHomebrew()
```
Comment (hatchan) [inline cmd/main.go]: We should probably log this error (as debug). If it isn't useful, then we probably do not want to return it from the function.

File: cmd/main.go [hunk 3]
```diff
@@ -770,6 +771,19 @@ func cmdVersion(options *core.VersionOptions) error {
...
+
+			if homebrew {
+				logger.Info("\nLooks like wercker was installed with homebrew.\n\n" +
+					"To update to the latest version please use:\n" +
+					"brew upgrade wercker-cli")
+				os.Exit(1)
```
Comment (hatchan) [inline cmd/main.go]: I would prefer having a "Are you sure you want to continue" here which defaults to `no`.

File: util/homebrew.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+	r := regexp.MustCompile(`(?m)^wercker-cli\s(.*)$`)
+	matches := r.FindAllStringSubmatch(version, -1)
+
+	if len(matches) == 0 {
+		// wercker not installed in homebrew
+		return false, nil
+	}
+
+	// resolve path to installed homebrew binary
+	cmd = exec.Command("brew", "--prefix")
```
Comment (hatchan) [inline util/homebrew.go]: Are we sure we want to check if the path is the same? Maybe just checking if brew it is installed is enough

**Comment by akupila** [comment]
Only parsing JSON might work but (as far as i could see) there's no info there about location of the binary. This tries to determine if the currently running binary is the one installed from homebrew. 

Of course making this simpler and just checking if homebrew is installed and if wercker was installed in homebrew might be enough. I guess it's _unlikely_ somebody would have installed wercker with homebrew but run a separately downloaded binary..

If this is the case we could simply get the JSON (should be stable according to brew docs) and look for wercker in there. If the brew is installed and wercker is found we'll assume you're using this version.

**Comment by akupila** [comment]
@hatchan @faiq Now much simpler check than before. Just tries to get info about wercker from homebrew and if that fails it assumes it's not installed with homebrew.

File: util/homebrew.go [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package util
+
+import "os/exec"
+
+// InstalledWithHomebrew tries to determine if the cli was installed with homebrew
+func InstalledWithHomebrew() bool {
+	// Check if homebrew is installed
+	cmd := exec.Command("brew", "info", "wercker-cli")
+	err := cmd.Start()
```
Comment (hatchan) [inline util/homebrew.go]: You need to use the `Run` instead of `Start`

**Comment by hatchan** [comment]
:+1:

**Comment by faiq** [comment]
nice and intrusive! 👍

**Pull request merged**